.github/workflows/publish.yaml

```diff
@@ -0,0 +1,97 @@
+name: publish
+on:
+  push:
+    tags:
+      - 'release-*'
+  workflow_dispatch:
+
+env:
+  NODE_VERSION : '20'
+jobs:
+  # Verify the code before publishing
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+
+      # Get the yarn cache path.
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "dir=$(yarn config get cacheFolder)" >> $GITHUB_OUTPUT
+      - name: Restore yarn cache
+        uses: actions/cache@v4
+        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: yarn-cache-folder-${{ hashFiles('**/yarn.lock', '.yarnrc.yml') }}
+          restore-keys: |
+            yarn-cache-folder-
+
+      - name: yarn install
+        uses: borales/actions-yarn@v5
+        with:
+          cmd: install --immutable
+      - name: yarn tsc
+        uses: borales/actions-yarn@v5
+        with:
+          cmd: tsc
+      - name: yarn build
+        uses: borales/actions-yarn@v5
+        with:
+          cmd: build:all
+
+  release:
+    needs: build
+
+    runs-on: ubuntu-latest
+
+    environment: publish-npm # This is github environment name
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+          registry-url: 'https://registry.npmjs.org'
+
+      # Get the yarn cache path.
+      - name: Get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "dir=$(yarn config get cacheFolder)" >> $GITHUB_OUTPUT
+      - name: Restore yarn cache
+        uses: actions/cache@v4
+        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: yarn-cache-folder-${{ hashFiles('**/yarn.lock', '.yarnrc.yml') }}
+          restore-keys: |
+            yarn-cache-folder-
+
+      - name: yarn install
+        uses: borales/actions-yarn@v5
```

Comment:
Per GitHub security recommendations, all (3rd-party) actions should be pinned to a commit hash because these tags are mutable.

https://github.com/github/codeql/blob/main/actions/ql/src/Security/CWE-829/UnpinnedActionsTag.md